Implement point sources

piyueh opened the pull request:
See the issue #2 for descriptions.

This pull request allows for setting up multiple point sources in `setrun.py`. The time histories of volumetric rates of the point sources accept piecewise constants.

An example is provided in `${GEOCLAW}/examples/landspill`. And new source files are in `${GEOCLAW}/src/2d/shallow/landspill`. The following modifications are done directly in GeoClaw original files:
1. `src/2d/shallow/amr2.f90`
2. `src/2d/shallow/src2.f90`
3. `src/python/geoclaw/data.py`

Not sure who's supposed to be the reviewer.

piyueh committed:
draft of point_source class

piyueh committed:
add I/O and cell_id to point_source class

piyueh committed:
draft of landspill module and point source I/O

- initialize landspill module and set_landspill
- python interface for specifying point source data in setrun.py
- use instrinsic function new_line in generic write
- add set_landspill to amr2.f90
- modify Makefile.geoclaw

piyueh committed:
incorporate point sources to src2.f90

- add a function to apply point sources
- fix a bug in cell_id subroutine of point source instances
- add function call for applying point sources to src2.f90

piyueh committed:
provide an example to test point sources

piyueh committed:
add a docker file to test PR8 in a clean environment

piyueh committed:
add a README for docker files

piyueh commented:
***Update***:
Add a docker file (`$GEOCLAW/docker/Dockerfile-PR8`) to create an image for testing.
Add a README (`$GEOCLAW/docker/README.md`) for how to use the image.

piyueh committed:
modify the interface of using point source objects

- a new class PointSourceCollection to encapsulate point sources
- rename class point_source -> PointSource
- rewrite I/O interface
- put point-source-related source files to a sub-folder

piyueh committed:
add a function to set trivial amount for IC to triger refinement

mandli commented on examples/landspill/Makefile:
```diff
@@ -0,0 +1,69 @@
+all: 
+	$(MAKE) topo
+	$(MAKE) .data
+	FFLAGS="-O2 -fopenmp" $(MAKE) .exe
```
Putting the compiler flags into one of the targets is not standard GeoClaw build behavior.  If you want to have default flags I recommend putting it in the `FFLAGS` line above so that someone could easily override these settings via the command line.

piyueh replied:
I agree. I'll change it.

piyueh replied:
Done in 61455d9.

mandli commented on examples/landspill/Makefile:
```diff
@@ -0,0 +1,69 @@
+	$(MAKE) topo
+	$(MAKE) .data
+	FFLAGS="-O2 -fopenmp" $(MAKE) .exe
+	$(MAKE) .output
```
`$(MAKE) .output` will also run `$(MAKE) .exe` if need be.  Did you want to explicitly run them separately?

piyueh replied:
Yes, you are right. I'll change it.

piyueh replied:
Done in 61455d9.

mandli commented on examples/landspill/download_topo.py:
```diff
@@ -0,0 +1,46 @@
+#! /usr/bin python3
```
Recommend using `#!/usr/bin/env python` instead to include greater compatibility.

piyueh replied:
What I am worried about is the use of `urllib.request` module with Python 2. I haven't fully tested the script with Python 2 environments. And I know some Linux systems use Python 2 as the default.

I'll test it with Python 2 and see if I can make it work if there's something incompatible with Python 2.

piyueh replied:
BTW, yes, I should use `/usr/bin/env` instead of `/usr/bin`. Using `/use/bin` is a bug.

piyueh replied:
Done in 61455d9.

mandli commented on src/2d/shallow/landspill/landspill_module.f90:
```diff
@@ -0,0 +1,35 @@
+    subroutine set_landspill(point_source_file)
+        character(len=*), intent(in), optional:: point_source_file 
+
+        ! open data file for point source collection
```
Make sure to set `module_setup = .true.` in this function (once it does something anyway).

piyueh replied:
Thanks! I'll work on this one.

piyueh replied:
Done in 61455d9.

mandli commented on src/2d/shallow/landspill/point_source/point_source_module.f90:
```diff
@@ -0,0 +1,296 @@
+        read(funit, *) this%t
+        read(funit, *) this%v_rate
+
+        ! TODO: make sure the t array is sorted and v_rate is sorted based on t
```
Just an FYI, there's a quicksort that is implemented already in AMRClaw that might be useful for sorting these.

piyueh replied:
Yes, I think I found it. If I'm correct, that should be the subroutine `qsorti` in `quick_sort1.f`. I'll work on this.

piyueh replied:
On second thought, I think if `t` array is not sorted, then it is an error of input data. So I believe checking whether a user-provided list is sorted or not should be done during the preprocessing stage. That is, the checking should happen when users do `make .data`.

So I will not do either sorting or checking here. I will do the checking in `data.py`, instead.

piyueh replied:
Done in 61455d9.

mandli commented on src/2d/shallow/landspill/point_source/point_source_module.f90:
```diff
@@ -0,0 +1,296 @@
+        character(*), intent(inout):: msg
+
+        ! code
+        write(*, *) "Direct read of this object is prohibited!"
```
I tend to prefer to explicitly use `print *,` if this is really intended to go to the console.

piyueh replied:
Yes. I think using `print *,` indeed provides a more clear hint that this is going to the console. I'll work on this one.

piyueh replied:
Done in 61455d9.

mandli commented on src/2d/shallow/qinit.f90:
```diff
@@ -26,6 +27,10 @@ subroutine qinit(meqn,mbc,mx,my,xlower,ylower,dx,dy,q,maux,aux)
     endif
 
+    ! add an almost trivial amount to the cells holding point sources
+    call point_sources%apply_trivial_ic(&
```
Is this required for stability reasons?  I might suggest keeping this below `dry_tolerance` rather than slightly above.  Also this may cause an arbitrary amount of initial fluid to be added to the domain depending on the size of the containing grid cell along with some weird behavior if you have multiple resolutions under the point source.

piyueh replied:
The reasons I added this are:
1. to trigger an initial adaptive refinement around point sources, and
2. for the solver to properly predict the time-step size of the second time-step.

For our cases, at the beginning, there is no working fluid in the domain. And I found the solver, therefore, fails to predict the proper time-step size for the second time-step in some cases. In these cases, the solver uses a very large time-step, that is likely the time gap between the current time and the time for the next result output.

I agree this workaround is not that justified. But I wasn't sure how to resolve the two abovementioned problems without touching much original GeoClaw code. And I think if the initial amounts added to the domain are close to the `dry_tolerance`, at least the error should not affect simulations.

I agree that it may be proper to keep the amount below `dry_tolerance`. I'm going to test it.

piyueh replied:
I'm thinking about other ways to resolve the problems.

piyueh replied:
See the update in [this comment](https://github.com/barbagroup/geoclaw/pull/8#issuecomment-430018402).

mandli commented on src/python/geoclaw/data.py:
```diff
@@ -529,3 +529,70 @@ def write(self, out_file='./multilayer.data', datasource="setrun.py"):
+        super(PointSourceData, self).__init__()
+
+        # number of point sources
+        self.add_attribute('n_point_sources', 0)
```
Do you need this?  Seems like `len(self.point_sources)` would always be equal to this.

piyueh replied:
It's just for preventing users from providing mismatched data in `setrun.py`. So the program is able to identify mistakes in the provided input data before a simulation starts. It's just like the concept of defensive driving. I'll prefer to keep it as this doesn't cause any serious performance overhead and provides an early-stage debugging layer.

mandli replied:
So you are trying to save the user from themselves?  In general I could see a user being frustrated by adding another point-source and then forgetting to increment `n_point_sources`.  If a user accidentally left out a point source I would hazard the guess that they would probably realize it especially considering I don't think you are going to assume that you have more than a very small number of point sources?

mandli reviewed (changes requested):
One more general suggestion - I might suggest more closely following the pattern we use for other GeoClaw examples, namely that examples are organized into broad application directories (e.g. tsunamis) and then contain possibly multiple specific examples (e.g. storm-surge/ike).

piyueh commented:
Regarding the naming of example directory, I agree with the suggestion. In fact, I am considering adding more example cases. So I will follow the GeoClaw convention.

piyueh committed:
modification based on the review of PR8

- add flexibility for FFLAGS in the Makefile of the example
- remove `make .data` and `make .exe` as `make .output` will do that
- add python 2 support to download_topo.py and fix the bug of
  `/usr/bin/env`
- set up module_setup to true after module initialization.
- replace `write(*, *)` with `print *,`
- check if the list of end times provided by users is sorted or not in
  `data.py`

piyueh committed:
move utah example from examples/landspill to examples/landspill/utah_dem

piyueh commented:
@mandli I modify the code for most of your comments in 61455d9. Also, the utah_dem example has been moved from `examples/landspill` to `examples/landspill/utah_dem` in 3a3e8a5. Thanks!

~The remaining issue is the one adding some trivial amount into the domain for point sources.~

**Update**:
In f5f787b, I modify the code to explicitly flag the cells containing point sources to be refined. So now no matter whether we add initial depth or not, the cells are always to be flagged when `t=t0`. And now the initial depths added to the point sources can be smaller than `dry_tolerance`. It's now `0.9*dry_tolerance`.

piyueh committed:
explicitly triger refinement around point sources

- explicitly triger refinement around point sources when t = t0
- reduce the initial depth at the location of the point sources ot
  0.9*dry_tolerance

piyueh committed:
change the max dt in the example

mandli commented:
So the remaining issues I believe can be better addressed by the following:

1.  Initial time step: You can set the initial time step but also initially taking a time step and finding that it violates the CFL condition will work as well and you will effectively compute the correct times step automatically.
2. Initial refinement:  Instead specifying a region around the point source for an initial time period.

piyueh committed:
remove the code adding trivial IC to the domain

piyueh commented:
Thanks, @mandli ! I removed the code that adds initial amounts to point source locations. We will handle that in future development if we encounter the initial time-step problem.

Regarding the initial refinement, I now just flag the cells where point sources are located in to be refined in `flag2refine2.f90` if `t=t0`. I think it works fine now. If there's any potential problem by doing so, please let me know! Thanks!

mandli commented:
I think the main issue is that it makes the `flag2refine` code here diverge from the regular one.  If you end up adding other specific flagging criteria this isn't that big of a deal as you would have to do this anyway but if not we may want to do this using a region.  We can of course revisit this later on and see what other criteria might be useful.
